File: server-java/src/main/java/com/vencedores/ski/service/VerificationService.java
package com.vencedores.ski.service;

import com.vencedores.ski.repository.DynamoDbRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import software.amazon.awssdk.services.dynamodb.model.AttributeValue;

import java.time.Instant;
import java.util.*;
import java.util.concurrent.ThreadLocalRandom;

@Service
public class VerificationService {

    private static final Logger log = LoggerFactory.getLogger(VerificationService.class);
    private static final int CODE_EXPIRY_SECONDS = 600; // 10 minutes
    private static final int MAX_ATTEMPTS = 5;
    private static final int MAX_SENDS = 3;

    private final DynamoDbRepository repo;
    private final EmailService emailService;

    public VerificationService(DynamoDbRepository repo, EmailService emailService) {
        this.repo = repo;
        this.emailService = emailService;
    }

    public void sendVerificationCode(String email) {
        String normalizedEmail = email.toLowerCase().trim();
        String pk = "VERIFY#" + normalizedEmail;
        String sk = "CODE";

        // Check rate limit
        var existing = repo.getItem(pk, sk);
        if (existing != null) {
            int sendCount = getNum(existing, "sendCount");
            long createdEpoch = Long.parseLong(str(existing, "createdEpoch"));
            long elapsed = Instant.now().getEpochSecond() - createdEpoch;
            if (elapsed < CODE_EXPIRY_SECONDS && sendCount >= MAX_SENDS) {
                log.info("[Verify] Rate limit reached for {}", normalizedEmail);
                return; // silently return to prevent enumeration
            }
        }

        String code = String.format("%06d", ThreadLocalRandom.current().nextInt(1_000_000));
        long now = Instant.now().getEpochSecond();

        var item = new HashMap<String, AttributeValue>();
        item.put("PK", s(pk));
        item.put("SK", s(sk));
        item.put("code", s(code));
        item.put("attempts", n(0));
        item.put("sendCount", n(existing != null ? getNum(existing, "sendCount") + 1 : 1));
        item.put("createdEpoch", s(String.valueOf(now)));
        item.put("expiresAt", n(now + CODE_EXPIRY_SECONDS));
        repo.putItem(item);

        emailService.sendVerificationCodeAsync(normalizedEmail, code);
        log.info("[Verify] Code sent to {}", normalizedEmail);
    }

    public Map<String, Object> verifyCode(String email, String code) {
        String normalizedEmail = email.toLowerCase().trim();
        String pk = "VERIFY#" + normalizedEmail;
        String sk = "CODE";

        var item = repo.getItem(pk, sk);
        if (item == null) {
            return Map.of("verified", false, "error", "invalid_code");
        }

        // Check expiry
        long expiresAt = Long.parseLong(item.get("expiresAt").n());
        if (Instant.now().getEpochSecond() > expiresAt) {
            repo.deleteItem(pk, sk);
            return Map.of("verified", false, "error", "code_expired");
        }

        // Check attempts
        int attempts = getNum(item, "attempts");
        if (attempts >= MAX_ATTEMPTS) {
            repo.deleteItem(pk, sk);
            return Map.of("verified", false, "error", "max_attempts");
        }

        String storedCode = str(item, "code");
        if (!storedCode.equals(code.trim())) {
            // Increment attempts
            repo.updateItem(pk, sk,
                    "SET attempts = attempts + :one",
                    Map.of(":one", AttributeValue.builder().n("1").build()),
                    null);
            return Map.of("verified", false, "error", "invalid_code");
        }

        // Success — delete verification item and return profile
        repo.deleteItem(pk, sk);

        var profile = getLatestProfile(normalizedEmail);
        if (profile == null) {
            return Map.of("verified", true, "profile", Map.of());
        }

        return Map.of("verified", true, "profile", profile);
    }

    private Map<String, Object> getLatestProfile(String email) {
        var items = repo.queryGsi("GSI2", "GSI2PK", "EMAIL#" + email, "GSI2SK", null);
        if (items.isEmpty()) {
            return null;
        }

        // Get the most recent registration (last item, sorted by GSI2SK = createdAt)
        var latest = items.getLast();

        var profile = new LinkedHashMap<String, Object>();
        profile.put("firstName", str(latest, "firstName"));
        profile.put("lastName", str(latest, "lastName"));
        profile.put("email", str(latest, "email"));
        profile.put("phone", str(latest, "phone"));
        profile.put("dob", str(latest, "dob"));
        profile.put("city", str(latest, "city"));
        profile.put("state", str(latest, "state"));
        profile.put("emergencyName", str(latest, "emergencyName"));
        profile.put("emergencyPhone", str(latest, "emergencyPhone"));
        profile.put("emergencyRelation", str(latest, "emergencyRelation"));
        profile.put("skillLevel", str(latest, "skillLevel"));
        profile.put("dietary", str(latest, "dietary"));
        profile.put("medConditions", str(latest, "medConditions"));
        profile.put("conditionDetails", str(latest, "conditionDetails"));
        profile.put("medAllergies", str(latest, "medAllergies"));
        profile.put("allergyDetails", str(latest, "allergyDetails"));
        profile.put("medMedications", str(latest, "medMedications"));
        profile.put("medicationDetails", str(latest, "medicationDetails"));
        return profile;
    }

    private String str(Map<String, AttributeValue> item, String key) {
        var v = item.get(key);
        return v != null && v.s() != null ? v.s() : "";
    }

    private int getNum(Map<String, AttributeValue> item, String key) {
        var v = item.get(key);
        return v != null && v.n() != null ? (int) Double.parseDouble(v.n()) : 0;
    }

    private AttributeValue s(String val) {
        return AttributeValue.builder().s(val != null ? val : "").build();
    }

    private AttributeValue n(long val) {
        return AttributeValue.builder().n(String.valueOf(val)).build();
    }
}

File: server-lambda/src/services/verification-service.ts
import * as repo from '../repository/dynamo-repository.js';
import * as emailService from './email-service.js';
import { isValidEmail } from '../utils/validation.js';

const CODE_EXPIRY_SECONDS = 600; // 10 minutes
const MAX_ATTEMPTS = 5;
const MAX_SENDS = 3;

export async function sendVerificationCode(email: string): Promise<void> {
  // Silently ignore malformed addresses (no send, no enumeration signal).
  if (!isValidEmail(email)) return;

  const normalizedEmail = email.toLowerCase().trim();
  const pk = `VERIFY#${normalizedEmail}`;
  const sk = 'CODE';

  // Check rate limit
  const existing = await repo.getItem(pk, sk);
  if (existing) {
    const sendCount = (existing.sendCount as number) || 0;
    const createdEpoch = parseInt(existing.createdEpoch as string, 10);
    const elapsed = Math.floor(Date.now() / 1000) - createdEpoch;
    if (elapsed < CODE_EXPIRY_SECONDS && sendCount >= MAX_SENDS) {
      console.info(`[Verify] Rate limit reached for ${normalizedEmail}`);
      return; // silently return to prevent enumeration
    }
  }

  const code = String(Math.floor(Math.random() * 1_000_000)).padStart(6, '0');
  const now = Math.floor(Date.now() / 1000);

  const item: Record<string, unknown> = {
    PK: pk,
    SK: sk,
    code,
    attempts: 0,
    sendCount: existing ? ((existing.sendCount as number) || 0) + 1 : 1,
    createdEpoch: String(now),
    expiresAt: now + CODE_EXPIRY_SECONDS,
  };
  await repo.putItem(item);

  await emailService.sendVerificationCode(normalizedEmail, code);
  console.info(`[Verify] Code sent to ${normalizedEmail}`);
}

export async function verifyCode(
  email: string,
  code: string
): Promise<Record<string, unknown>> {
  const normalizedEmail = email.toLowerCase().trim();
  const pk = `VERIFY#${normalizedEmail}`;
  const sk = 'CODE';

  const item = await repo.getItem(pk, sk);
  if (!item) {
    return { verified: false, error: 'invalid_code' };
  }

  // Check expiry
  const expiresAt = item.expiresAt as number;
  if (Math.floor(Date.now() / 1000) > expiresAt) {
    await repo.deleteItem(pk, sk);
    return { verified: false, error: 'code_expired' };
  }

  // Check attempts
  const attempts = (item.attempts as number) || 0;
  if (attempts >= MAX_ATTEMPTS) {
    await repo.deleteItem(pk, sk);
    return { verified: false, error: 'max_attempts' };
  }

  const storedCode = item.code as string;
  if (storedCode !== code.trim()) {
    // Increment attempts
    await repo.updateItem(pk, sk, 'SET attempts = attempts + :one', { ':one': 1 }, null);
    return { verified: false, error: 'invalid_code' };
  }

  // Success — delete verification item and return profile
  await repo.deleteItem(pk, sk);

  const profileResult = await getLatestProfile(normalizedEmail);
  if (!profileResult) {
    // Code was valid, but this email has no registration on file — signal that
    // explicitly so the client doesn't drop the user into a blank "returning"
    // form (e.g. they registered with a different email).
    return { verified: true, profile: null, registeredEventIds: [], registrations: [] };
  }

  return {
    verified: true,
    profile: profileResult.profile,
    registeredEventIds: profileResult.registeredEventIds,
    registrations: profileResult.registrations,
  };
}

async function getLatestProfile(email: string) {
  const items = await repo.queryGsi('GSI2', 'GSI2PK', `EMAIL#${email}`, 'GSI2SK', null);
  if (items.length === 0) return null;

  const registeredEventIds = [
    ...new Set(
      items
        .map((item) => item.eventId as string)
        .filter((id) => id && id.length > 0)
    ),
  ];

  // Minors are registered under their guardian's email but are not the account
  // holder — exclude them so the prefilled profile and balance lookups use the
  // adult's own registration.
  const adultItems = items.filter((item) => item.isMinor !== true);

  // Each registration represents the guardian's whole group: the guardian plus
  // any minors linked to them. Totals are summed across the group so the
  // returning screen can show one balance and one participant list.
  const registrations = adultItems
    .filter((item) => (item.eventId as string)?.length > 0)
    .map((guardian) => {
      const guardianId = guardian.id as string;
      const minorItems = items.filter(
        (it) => it.isMinor === true && (it.guardianRegId as string) === guardianId
      );
      const group = [guardian, ...minorItems];
      const participants = group.map((p) => ({
        firstName: (p.firstName as string) ?? '',
        lastName: (p.lastName as string) ?? '',
        isMinor: p.isMinor === true,
      }));
      const totalPaid = group.reduce((sum, p) => sum + ((p.totalPaid as number) || 0), 0);
      const totalOwed = group.reduce((sum, p) => sum + ((p.totalOwed as number) || 0), 0);
      return {
        eventId: guardian.eventId ?? '',
        confirmationId: guardianId,
        totalPaid,
        totalOwed,
        paymentStatus: guardian.paymentStatus ?? '',
        participants,
      };
    });

  // Get the most recent adult registration (last item, sorted by GSI2SK = createdAt)
  const profileSource = adultItems.length > 0 ? adultItems : items;
  const latest = profileSource[profileSource.length - 1];

  const profile = {
    firstName: latest.firstName ?? '',
    lastName: latest.lastName ?? '',
    email: latest.email ?? '',
    phone: latest.phone ?? '',
    dob: latest.dob ?? '',
    emergencyName: latest.emergencyName ?? '',
    emergencyPhone: latest.emergencyPhone ?? '',
    emergencyRelation: latest.emergencyRelation ?? '',
    skillLevel: latest.skillLevel ?? '',
    dietary: latest.dietary ?? '',
    medConditions: latest.medConditions ?? '',
    conditionDetails: latest.conditionDetails ?? '',
    medAllergies: latest.medAllergies ?? '',
    allergyDetails: latest.allergyDetails ?? '',
    medMedications: latest.medMedications ?? '',
    medicationDetails: latest.medicationDetails ?? '',
  };

  return { profile, registeredEventIds, registrations };
}
